src/main/java/io/cdap/plugin/switchcase/route/BasicPortSpecificationEvaluator.java

```diff
@@ -53,24 +54,76 @@
-    if (value == null) {
-      LOG.debug("Found null value for {}.", routingField);
-      throw new PortNotSpecifiedException(PortNotSpecifiedException.Reason.NULL);
-    }
-    String textValue = String.valueOf(value);
+    Object value = getRoutingFieldValue(record);
+    Schema schema = getRoutingFieldSchema(record);
     for (BasicPortSpecification portSpecification : portSpecifications) {
       String portName = portSpecification.getName();
       BasicRoutingFunction routingFunction = portSpecification.getRoutingFunction();
-      if (routingFunction.evaluate(textValue, portSpecification.getParameter())) {
+      if (routingFunction.evaluate(value, portSpecification.getParameter(), schema)) {
         return portName;
       }
     }
 
     throw new PortNotSpecifiedException(PortNotSpecifiedException.Reason.DEFAULT);
   }
 
+  private Schema getRoutingFieldSchema(StructuredRecord record) {
+    Schema.Field field = record.getSchema().getField(routingField);
+    if (field == null) {
+      // this should never happen, since validation ensures that the routing field exists in the schema
+      throw new IllegalArgumentException(String.format("Routing field %s not found in input schema", routingField));
+    }
+    return field.getSchema();
+  }
+
+  private Object getRoutingFieldValue(StructuredRecord record) throws PortNotSpecifiedException {
+    Object returnValue = record.get(routingField);
+    if (returnValue == null) {
+      LOG.debug("Found null value for routing field {}.", routingField);
+      throw new PortNotSpecifiedException(PortNotSpecifiedException.Reason.NULL);
+    }
+    Schema schema = getRoutingFieldSchema(record);
+    Schema.Type type = schema.isNullable() ? schema.getNonNullable().getType() : schema.getType();
+    switch (type) {
+      case STRING:
+      case FLOAT:
+      case DOUBLE:
+        return returnValue;
+      case INT:
+      case LONG:
+      case BYTES:
+        // INT can be INT, DATE, TIME_MILLIS
+        // LONG can be LONG, TIMESTAMP_MILLIS, TIMESTAMP_MACROS, TIME_MACROS
+        // BYTES can be decimal
+        Schema.LogicalType logicalType = schema.getLogicalType();
+        if (logicalType == null) {
+          return returnValue;
+        }
+        return getLogicalValue(record, logicalType);
+      default:
+        // this should never happen since validation already validates the data types
+        throw new IllegalArgumentException(
+          String.format("Unsupported data type %s for routing field %s", type, routingField)
+        );
+    }
+  }
+
+  private Object getLogicalValue(StructuredRecord record, Schema.LogicalType logicalType) {
+    switch (logicalType) {
+      case DATE:
+        return record.getDate(routingField);
+      case TIME_MICROS:
+      case TIME_MILLIS:
+        return record.getTime(routingField);
+      case TIMESTAMP_MICROS:
+      case TIMESTAMP_MILLIS:
+        return record.getTimestamp(routingField);
+      // TODO: support decimal type?
```

Comment:
why decimal type is not supported?

Reply:
good catch. This is a mistake. Pushed another commit that fixes this, and a few other things (datatypes in the test were wrong)